File: check_overlap.m
function [successflag,coords] = check_overlap(particlelist,params)
successflag=0;
%
population=particlelist.Nprime;
newind=population+1;
ac=particlelist.ac(newind,:);
q=particlelist.quat(newind,:);
boundingradius=max(ac);
newpolyhedra_unshifted=discretize_ellipsoid(ac,q,[0 0 0]);
%
maxattempts=10000;
perturbfreq=100; %resample new coords at this rate
maxradii=max(particlelist.ac(1:population,:),[],2);
pos=particlelist.xyz(1:population,:);

EPAclearance=0.1;
for numattempts = 1:maxattempts
    if mod(numattempts,perturbfreq)==1
        %this assumes that previous attempts at shifting coords were not too successful
        %sample new coords
        if numattempts<perturbfreq %first sampling is uniformly random
            coords=(rand(1,3)-0.5)*params.boxlength;
        else
            randvars=rand(1,3);
            zbins = (-5:5)/10*params.boxlength;
            znum = histcounts(pos(:,3),zbins);
            maxznum=max(znum);
            minznum=min(znum);
            if maxznum==minznum
                zcoord = (randvars(1)-0.5)*params.boxlength;
            else
                newznum = (maxznum-znum)/(maxznum-minznum);
                cumznum=[0 cumsum(newznum)]+(0:10)*1e-10; %Small constants added to make them unique
                cumznum=cumznum/cumznum(end);
                zcoord = interp1(cumznum,zbins,randvars(3));
            end
            zbinwidth=zbins(2)-zbins(1);
            zfilter=(pos(:,3)>zcoord-zbinwidth/2) & (pos(:,3)<=zcoord+zbinwidth/2);
            ybins = (-5:5)/10*params.boxlength;
            ynum = histcounts(pos(zfilter,2),ybins);
            maxynum=max(ynum);
            minynum=min(ynum);
            if ~sum(zfilter) || maxynum==minynum
                ycoord = (randvars(2)-0.5)*params.boxlength;
            else
                newynum = (maxynum-ynum)/(maxynum-minynum);
                cumynum=[0 cumsum(newynum)]+(0:10)*1e-10; %Small constants added to make them unique
                cumynum=cumynum/cumynum(end);
                ycoord = interp1(cumynum,ybins,randvars(2));
            end
            ybinwidth=ybins(2)-ybins(1);
            yzfilter=(pos(zfilter,2)>ycoord-ybinwidth/2) & (pos(zfilter,2)<=ycoord+ybinwidth/2);
            xbins = (-5:5)/10*params.boxlength;
            xnum = histcounts(pos(yzfilter,1),xbins);
            maxxnum=max(xnum);
            minxnum=min(xnum);
            if ~sum(yzfilter) || maxxnum==minxnum
                xcoord = (randvars(2)-0.5)*params.boxlength;
            else
                newxnum = (maxxnum-xnum)/(maxxnum-minxnum);
                cumxnum=[0 cumsum(newxnum)]+(0:10)*1e-10; %Small constants added to make them unique
                cumxnum=cumxnum/cumxnum(end);
                xcoord = interp1(cumxnum,xbins,randvars(1));
            end
            coords=[xcoord ycoord zcoord];
        end
    end
    newpolyhedra=newpolyhedra_unshifted;
    newpolyhedra.vertices=newpolyhedra.vertices+ones(size(newpolyhedra.vertices,1),1)*coords;
    dist=sqrt(sum((ones(population,1)*coords-pos).^2,2));
    potentialoverlap=(maxradii+boundingradius-dist>0);
    if all(~potentialoverlap)
        successflag=1;
        break;
    end
    indices=find(potentialoverlap);
    numintersections=0;
    cummulativeshift=[0 0 0];
    for i=1:length(indices)
        [intersectionflag,intersectionsimplex] = gjk_simplex(newpolyhedra,particlelist.polyhedra(indices(i)));
        if intersectionflag
            numintersections=numintersections+1;
            [shiftvector,shiftdist] = expandingpolytope_shift(newpolyhedra,particlelist.polyhedra(indices(i)),intersectionsimplex);
            cummulativeshift = -shiftvector*(shiftdist+EPAclearance);
        end
    end
    if any(cummulativeshift)
        if ~mod(numattempts,100)
            disp(['Warning: Previous attempts failed to converge after ' num2str(numattempts) ' attemps. Current ellipsoid has ' num2str(numintersections) ' intersections. The population is ' num2str(population) '.']);
        end
        coords=coords+cummulativeshift;
        coords=mod(coords+params.boxlength/2,params.boxlength)-params.boxlength/2;
    else
        successflag=1;
        break;
    end
end
end

File: python_code/check_overlap.py
import numpy as np
from discretize_ellipsoid import discretize_ellipsoid
from gjk_simplex import gjk_simplex
from expandingpolytope_shift import expandingpolytope_shift

def check_overlap(particlelist, params):
    successflag = 0

    population = particlelist['Nprime']
    newind = population + 1
    ac = particlelist['ac'][newind - 1]
    q = particlelist['quat'][newind - 1]
    boundingradius = np.max(ac)
    newpolyhedra_unshifted = discretize_ellipsoid(ac, q, [0, 0, 0])

    maxattempts = 10000
    perturbfreq = 100  # resample new coords at this rate
    maxradii = np.max(particlelist['ac'][:population], axis=1)
    pos = particlelist['xyz'][:population]

    EPAclearance = 0.1
    for numattempts in range(1, maxattempts + 1):
        if numattempts % perturbfreq == 1:
            if numattempts < perturbfreq:
                coords = (np.random.rand(3) - 0.5) * params['boxlength']
            else:
                randvars = np.random.rand(3)
                zbins = np.linspace(-0.5 * params['boxlength'], 0.5 * params['boxlength'], 11)
                znum, _ = np.histogram(pos[:, 2], bins=zbins)
                maxznum = np.max(znum)
                minznum = np.min(znum)
                if maxznum == minznum:
                    zcoord = (randvars[0] - 0.5) * params['boxlength']
                else:
                    newznum = (maxznum - znum) / (maxznum - minznum)
                    cumznum = np.concatenate(([0],np.cumsum(newznum))) + np.arange(11) * 1e-10
                    cumznum /= cumznum[-1]
                    zcoord = np.interp(randvars[2], cumznum, zbins)
                zbinwidth = zbins[1] - zbins[0]
                zfilter = (pos[:, 2] > zcoord - zbinwidth / 2) & (pos[:, 2] <= zcoord + zbinwidth / 2)

                ybins = np.linspace(-0.5 * params['boxlength'], 0.5 * params['boxlength'], 11)
                ynum, _ = np.histogram(pos[zfilter, 1], bins=ybins)
                maxynum = np.max(ynum)
                minynum = np.min(ynum)
                if not np.sum(zfilter) or maxynum == minynum:
                    ycoord = (randvars[1] - 0.5) * params['boxlength']
                else:
                    newynum = (maxynum - ynum) / (maxynum - minynum)
                    cumynum = np.concatenate(([0],np.cumsum(newynum))) + np.arange(11) * 1e-10
                    cumynum /= cumynum[-1]
                    ycoord = np.interp(randvars[1], cumynum, ybins)
                ybinwidth = ybins[1] - ybins[0]
                yzfilter = zfilter & (pos[:, 1] > ycoord - ybinwidth / 2) & (pos[:, 1] <= ycoord + ybinwidth / 2)

                xbins = np.linspace(-0.5 * params['boxlength'], 0.5 * params['boxlength'], 11)
                xnum, _ = np.histogram(pos[yzfilter, 0], bins=xbins)
                maxxnum = np.max(xnum)
                minxnum = np.min(xnum)
                if not np.sum(yzfilter) or maxxnum == minxnum:
                    xcoord = (randvars[0] - 0.5) * params['boxlength']
                else:
                    newxnum = (maxxnum - xnum) / (maxxnum - minxnum)
                    cumxnum = np.concatenate(([0],np.cumsum(newxnum))) + np.arange(11) * 1e-10
                    cumxnum /= cumxnum[-1]
                    xcoord = np.interp(randvars[0], cumxnum, xbins)
                coords = np.array([xcoord, ycoord, zcoord])

        newpolyhedra = newpolyhedra_unshifted.copy()
        newpolyhedra['vertices'] = newpolyhedra['vertices'] + np.ones((newpolyhedra['vertices'].shape[0], 1)) * coords
        dist = np.sqrt(np.sum((np.ones((population, 1)) * coords - pos) ** 2, axis=1))
        potentialoverlap = (maxradii + boundingradius - dist > 0)

        if not np.any(potentialoverlap):
            successflag = 1
            break

        indices = np.where(potentialoverlap)[0]
        numintersections = 0
        cummulativeshift = np.array([0, 0, 0])
        for i in indices:
            #try:
            intersectionflag, intersectionsimplex = gjk_simplex(newpolyhedra, particlelist['polyhedra'][i])
            #except KeyError:
            #    continue

            if intersectionflag:
                numintersections += 1
                shiftvector, shiftdist = expandingpolytope_shift(newpolyhedra, particlelist['polyhedra'][i],
                                                                 intersectionsimplex)
                cummulativeshift = -shiftvector * (shiftdist + EPAclearance)

        if np.any(cummulativeshift):
            if not numattempts % 100:
                print(
                    f"Warning: Previous attempts failed to converge after {numattempts} attempts. Current ellipsoid has {numintersections} intersections. The population is {population}.")
            coords = coords + cummulativeshift
            coords = np.mod(coords + params['boxlength'] / 2, params['boxlength']) - params['boxlength'] / 2
        else:
            successflag = 1
            break

    return successflag, coords

